### PR TITLE
HPCC-22156 Secure socket readtms doesn't throw an exception

### PR DESCRIPTION
**Pull request by mayx**
- Throw JSOCKETErr_graceful_close if min_size>0 and SSL_read's return
  value is less or equal to 0

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by mayx** [comment]
@jakesmith @mckellyln please review...

**Comment by mayx** [comment]
@jakesmith @mckellyln just made a change, please review...

File: system/security/securesocket/securesocket.cpp [hunk 1]
```diff
@@ -774,27 +774,20 @@ void CSecureSocket::readTimeout(void* buf, size32_t min_size, size32_t max_size,
...
-                if(m_loglevel >= SSLogMax)
-                {
-                    char errbuf[512];
-                    ERR_error_string_n(ERR_get_error(), errbuf, 512);
-                    DBGLOG("Warning: SSL_read error %d - %s", err, errbuf);
-                }
-            }
-            break;
+            char errbuf[512];
+            ERR_error_string_n(err, errbuf, 512);
```
Comment (mckellyln) [inline system/security/securesocket/securesocket.cpp]: Why err and not ERR_get_error() here ?

Comment (mckellyln) [reply system/security/securesocket/securesocket.cpp]: maybe also call SSL_shutdown() ?

File: system/security/securesocket/securesocket.cpp [hunk 2]
```diff
@@ -774,27 +774,20 @@ void CSecureSocket::readTimeout(void* buf, size32_t min_size, size32_t max_size,
         }
 
         rc = SSL_read(m_ssl, (char*)buf + size_read, max_size - size_read);
```
Comment (mckellyln) [inline system/security/securesocket/securesocket.cpp]: Maybe call ERR_clear_error() before SSL_read() ?

File: system/security/securesocket/securesocket.cpp (hunk 1 above)
Comment (mayx) [reply system/security/securesocket/securesocket.cpp]: ERR_get_error() returns the same error code, with the effect of taking the error off the stack. ERR_clear_error() does the same and will clear up everything.

File: system/security/securesocket/securesocket.cpp (hunk 2 above)
Comment (mayx) [reply system/security/securesocket/securesocket.cpp]: Yes ERR_clear_error() should be called before and after. I was thinking not to change the logic for now but do it in a separate issue (a few other places could be added), but I can add it here now.

File: system/security/securesocket/securesocket.cpp [hunk 3]
```diff
@@ -774,27 +774,20 @@ void CSecureSocket::readTimeout(void* buf, size32_t min_size, size32_t max_size,
...
-                }
-            }
-            break;
+            char errbuf[512];
+            ERR_error_string_n(err, errbuf, 512);
+            ERR_clear_error();
+            if (m_loglevel >= SSLogMax)
+                DBGLOG("Warning: SSL_read error %d - %s", err, errbuf);
+            if (min_size > 0)
+                throw MakeStringException(-1, "SSL_read error %d - %s", err, errbuf);
```
[the remark this answers is not in the document]
Comment (mayx) [reply system/security/securesocket/securesocket.cpp]: I will change it to JSOCKERR_graeful_close.

File: system/security/securesocket/securesocket.cpp (hunk 1 above)
Comment (mayx) [reply system/security/securesocket/securesocket.cpp]: Don't believe we should call SSL_shutdown() here. Manual explicitly says not to call for certain errors.

**Comment by mayx** [comment]
@jakesmith @mckellyln change made per suggestions, please review again...

**Comment by jakesmith** [comment]
Looks good, didn't spot any issues.
@mckellyln ?

File: system/security/securesocket/securesocket.cpp [hunk 4]
```diff
@@ -773,28 +773,23 @@ void CSecureSocket::readTimeout(void* buf, size32_t min_size, size32_t max_size,
...
-            }
-            break;
+            char errbuf[512];
+            ERR_error_string_n(err, errbuf, 512);
+            ERR_clear_error();
+            VStringBuffer errmsg("SSL_read error %d - %s", err, errbuf);
+            if (m_loglevel >= SSLogMax)
+                DBGLOG("Warning: %s", errmsg.str());
+            if (min_size > 0)
+                throw createJSocketException(JSOCKERR_graceful_close, errmsg);
```
Comment (mckellyln) [inline system/security/securesocket/securesocket.cpp]: errmsg.str() ?  Does it matter ?

Comment (mayx) [reply system/security/securesocket/securesocket.cpp]: Don't think it matters thanks to the const char* operator overloading

Comment (mckellyln) [reply system/security/securesocket/securesocket.cpp]: ok, then looks fine

**Review by mckellyln — APPROVED** [review]
@jakesmith @mayx approved

File: system/security/securesocket/securesocket.cpp (hunk 4 above)
Comment (mayx) [reply system/security/securesocket/securesocket.cpp]: Had to add .str() to DBGLOG because parameter is char* not const char*

**Comment by jakesmith** [comment]
@richardkchapman - ready to merge.

**Pull request merged**